kube-proxy refactoring

danwinship opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind bug

#### What this PR does / why we need it:
Refactors kube-proxy's internals to reduce duplicated code and make it easier to add new backends. See https://docs.google.com/document/d/10_bpV_-IAOWO3DGRH-qrQBNySmqFKbTHUcw_tgvuAa8/edit#

I was planning to submit this in pieces but @aojea wanted to see the whole thing all at once.

At the moment, everything up to "Rework kube-proxy single/dual-stack proxier creation" is stable-ish, the commits after that are more in flux. ~The first three commits are also in #117297~

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/sig network
/priority important-longterm
/triage accepted
/assign @aojea

danwinship committed:
Rework kube-proxy single/dual-stack proxier creation

Add a supportedIPFamilies() method to ProxyServer and have the
platform-independent code use that to decide whether it will run as
single-stack IPv4, single-stack IPv6, or dual-stack. (Kube-proxy does
not need to distinguish IPv4-primary dual-stack from IPv6-primary
dual-stack.)

Change createProxier() to create a single Proxier of a given IP
family, and have the platform-independent code call it once or twice
as needed, and then call metaproxy.NewMetaProxier() itself in the
dual-stack case.

danwinship committed:
Rename MetaProxier to Runner and use it for both single and dual stack

danwinship committed:
Start moving sync handling from proxy.Provider to proxy.Runner

danwinship committed:
Finish moving sync/healthz handling from proxy.Provider to proxy.Runner

danwinship committed:
Move informer handling from cmd/kube-proxy to proxy.Runner

danwinship committed:
Move informer sync / initialized handling into proxy.Runner

danwinship committed:
WIP Move all change tracking from proxy.Provider to proxy.Runner

danwinship committed:
Simplify getDualStackLocalDetectorTuple

Since the single-stack and dual-stack local-detector-getters now have
the same behavior in terms of error-checking and dual-stack config, we
can just replace the contents of getDualStackLocalDetectorTuple() with
a pair of calls to getLocalDetector().

MikeZappa87 commented on pkg/proxy/winkernel/proxier.go:
```diff
@@ -26,29 +26,22 @@ import (
-	"sync/atomic"
 	"time"
 
 	"github.com/Microsoft/hcsshim"
```
@princepereira just as an FYI

danwinship committed:
Remove unnecessary utiliptables.Interface arg from local detectors

getLocalDetector() used to pass a utiliptables.Interface to
NewDetectLocalByCIDR() so that NewDetectLocalByCIDR() could verify
that the passed-in CIDR was of the same family as the iptables
interface. It would make more sense for getLocalDetector() to verify
this itself and just *not call NewDetectLocalByCIDR* if the families
don't match, and that's what the code does now. So there's no longer
any need to pass the utiliptables.Interface to the local detector.

danwinship committed:
Improve the single-stack LocalDetector behavior

1. When bringing up a single-stack kube-proxy in a dual-stack cluster,
allow using either the primary or secondary IP family.

2. Since the earlier config-checking code will already have bailed out
if the single-stack configuration is unusably broken, we don't need to
do that here. Instead, just return a no-op local detector if there are
no usable CIDRs of the expected IP family.

danwinship committed:
Add names to the getLocalDetector unit tests, use t.Run

dims commented:
This work-in-progress PR needs a rebase. please rebase if this PR is still needed for the upcoming release.

danwinship commented:
(I'm still working on this bit by bit (eg, #123005 had some commits from here) but this particular PR was just an overview and not intended to merge as itself anyway.)
